geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/IndexKeySpace.scala

```diff
@@ -59,4 +63,99 @@ trait IndexKeySpace[T] {
     * @return
     */
   def getIndexValues(sft: SimpleFeatureType, filter: Filter, explain: Explainer): T
+
+  /**
+    * Creates ranges over the index keys
+    *
+    * @param values index values @see getIndexValues
+    * @return
+    */
+  def getRanges(values: T): Iterator[ScanRange[U]]
+
+  /**
+    * Creates bytes from ranges
+    *
+    * @param ranges typed scan ranges. @see `getRanges`
+    * @param prefixes prefixes to the range, if any. each prefix will create a new range
+    * @return
+    */
+  def getRangeBytes(ranges: Iterator[ScanRange[U]],
+                    prefixes: Seq[Array[Byte]] = Seq.empty,
+                    tier: Boolean = false): Iterator[ByteRange]
+
+  /**
+    * Determines if the ranges generated by `getRanges` are sufficient to fulfill the query,
+    * or if additional filtering needs to be done
+    *
+    * @param config data store config
+    * @param values index values @see getIndexValues
+    * @param hints query hints
+    * @return
+    */
+  def useFullFilter(values: Option[T],
+                    config: Option[GeoMesaDataStoreConfig],
+                    hints: Hints): Boolean
+}
+
+object IndexKeySpace {
+
+  type ToIndexKeyBytes = (Seq[Array[Byte]], SimpleFeature, Array[Byte]) => Seq[Array[Byte]]
+
+  sealed trait ByteRange
+
+  case class BoundedByteRange(lower: Array[Byte], upper: Array[Byte]) extends ByteRange
+  case class SingleRowByteRange(row: Array[Byte]) extends ByteRange
+  case class TieredByteRange(lower: Array[Byte],
+                             upper: Array[Byte],
+                             lowerTierable: Boolean = false,
+                             upperTierable: Boolean = false) extends ByteRange
+
+  object ByteRange {
+
+    import org.locationtech.geomesa.utils.conversions.ScalaImplicits.RichTraversableOnce
+
+    implicit val ByteOrdering: Ordering[Array[Byte]] =
+      Ordering.comparatorToOrdering(UnsignedBytes.lexicographicalComparator)
+
+    val UnboundedLowerRange: Array[Byte] = Array.empty
+    val UnboundedUpperRange: Array[Byte] = Array.fill(3)(ByteArrays.MaxByte)
```

Comment:
(Sanity checks): Is the empty Array of bytes the smallest possible Array[Byte]?  

On the other end, what would have to happen to have the first three bytes of a GeoMesa key be MaxByte?  If we had 255 shards.... and ...  (I suppose I'm wondering how often such a corner case could come up...)

Reply:
yeah, an empty array sorts before any non-empty array, so it should be good as a lower range. for the upper range, it doesn't involve shards or table sharing prefixes, so it's just the raw value. for our z-indices, we don't use the first 1-2 bits of the z value, so this should always sort after any valid value (z3 uses 2 bytes for the bin, then the z value, which serendipitously lines up with the 3 byte size i randomly chose). for attribute indices, it's possible but seems unlikely, since it's not a printable char...

Reply:
For fun, could you Z3-decode the 3 MaxBytes;)?

Reply:
hmm, let me try encoding the max z3 instead. The max bytes won't really work as a z-value, as i think having the sign bit set causes math errors. and the first thing it does when decoding is to apply a bitmask: https://github.com/locationtech/sfcurve/blob/master/zorder/src/main/scala/org/locationtech/sfcurve/zorder/Z3.scala#L83

Reply:
here's z3:
```
scala> val sfc = Z3SFC(TimePeriod.Week)
sfc: org.locationtech.geomesa.curve.Z3SFC = org.locationtech.geomesa.curve.Z3SFC@6fe595dc
scala> val z = sfc.index(sfc.lon.max, sfc.lat.max, sfc.time.max.toLong).z
z: Long = 9223372036854775807
scala> ByteArrays.toBytes(Short.MaxValue, z).map(ByteArrays.toHex).mkString(":")
res3: String = 7f:ff:7f:ff:ff:ff:ff:ff:ff:ff
scala> ByteArrays.toHex(ByteArrays.MaxByte)
res4: String = ff
```

Reply:
and here's z2:
```
scala> val sfc = org.locationtech.geomesa.curve.Z2SFC
sfc: org.locationtech.geomesa.curve.Z2SFC.type = org.locationtech.geomesa.curve.Z2SFC$@76ff68c5
scala> val z = sfc.index(sfc.lon.max, sfc.lat.max).z
z: Long = 4611686018427387903
scala> ByteArrays.toBytes(z).map(ByteArrays.toHex).mkString(":")
res0: String = 3f:ff:ff:ff:ff:ff:ff:ff
```